docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,209 @@
+
+  It was pointed out that the existing `providerIDList` suffers the same scalability issue as an `infrastructureRefList` would. It would be nice to refactor it away, but doing so is not considered a goal of this proposal effort.
+
+- The logic to convey the intent of replicas in MachinePool (and MachineDeployment) is currently provider-agnostic and driven by the core controllers. The approach in this proposal inverts that, delegating the responsibility to the provider first to create the infrastructure resources, and then having the CAPI core controllers sync up with those resources by creating and deleting Machines.
+
+  One goal of creating these resources first in the provider code is to allow responsiveness to a MachinePool's native resource. A provider's MachinePool controller can keep the infrastructure representation of the actual scalable resource more current, perhaps responding to native platform events.
+
+  There is a potential that this approach creates a bad representation of intent with regard to MachinePool Machines depending on the provider's implementation. We assume that the number of implementors is small, and that the providers are well-known. A "reference implementation" of MachinePool Machines for Docker should guide providers towards reliable code.
+
+## Alternatives
+
+We could consider refactoring MachineDeployment/MachineSet to support the option of using cloud infrastructure-native scaling group resources. That is, merge MachinePool into MachineDeployment. MachinePool and MachineDeployment do overlap significantly, and this proposal aims to eliminate one of their major differences: that individual Machines aren’t represented in a MachinePool.
+
+While it is valuable that MachineDeployments are provider-agnostic, MachinePools take advantage of scalable resources unique to an infra provider, which may have advantages in speed or reliability. MachinePools allow an infrastructure provider to decide which native features to use, while still conforming to some basic common behaviors.
+
+ To merge MachinePool behavior — along with the changes proposed here — into MachineDeployment would effectively deprecate the experimental MachinePool API and would increase the overall scope of work significantly, while increasing challenges to maintaining backward compatibility in the stable API.
+
+## Upgrade Strategy
+
+MachinePool Machines will be a backward-compatible feature. Existing infrastructure providers can make no changes and will observe the same behavior with MachinePools they always have.
+
+## Additional Details
+
+### clusterctl client
+
+Currently, the clusterctl discovery service searches for MachineDeployments and MachineSets and can list the Machines in MachineDeployments and MachineSets but does not search for MachinePools. The clusterctl client will then be updated to discover MachinePools, MachinePool Machines, as well as the associated BootstrapConfig and MachineInfrastructure objects.
```

Comment:
isn't this already done from @Jont828's discovery PR?

Reply:
Not quite, in the discovery PR we can show the MachinePool infrastrucutre and BootstrapConfig but not the MachinePoolMachines or their associated resources. I can update the proposal to say that we can discover MPs with the discovery change though.